implementation/src/main/java/io/smallrye/config/AbstractLocationConfigSourceFactory.java

```diff
@@ -0,0 +1,247 @@
+
+/**
+ * This {@code AbstractLocationConfigSourceFactory} allows to initialize additional config locations with the
+ * configuration {@link AbstractLocationConfigSourceFactory#SMALLRYE_LOCATIONS}. The configuration support multiple
+ * locations separated by a comma and each much represent a valid {@link URI}.
+ * <p>
+ *
+ * The locations comprise a list of URIs which are loaded in order. The following URI schemes are supported:
+ *
+ * <ol>
+ * <li>file or directory</li>
+ * <li>classpath resource</li>
+ * <li>jar resource</li>
+ * <li>http resource</li>
+ * </ol>
+ * <p>
+ *
+ * If a profile is active, the profile resource is only loaded if the unprofiled resource is available in the same
+ * location. This is to keep a consistent loading order and matching with the unprofiled resource. Profiles are not
+ * taken into account if the location is a directory.
+ */
+@Experimental("Load additional config locations")
+public abstract class AbstractLocationConfigSourceFactory implements ConfigSourceFactory {
+    public static final String SMALLRYE_LOCATIONS = "smallrye.config.locations";
+
+    protected abstract String[] getFileExtensions();
+
+    protected abstract ConfigSource loadConfigSource(final URL url, final int ordinal) throws IOException;
+
+    protected ConfigSource loadConfigSource(final URL url) throws IOException {
+        return this.loadConfigSource(url, ConfigSource.DEFAULT_ORDINAL);
+    }
+
+    @Override
+    public Iterable<ConfigSource> getConfigSources(final ConfigSourceContext context) {
+        final ConfigValue value = context.getValue(SMALLRYE_LOCATIONS);
+
+        final List<ConfigSource> configSources = new ArrayList<>();
+        if (value.getValue() != null) {
+            List<URI> resources = newCollectionConverter(new URIConverter(), ArrayList::new).convert(value.getValue());
+            if (resources != null) {
+                for (URI uri : resources) {
+                    if (uri.getScheme() == null || uri.getScheme().equals("file")) {
+                        tryFileSystem(uri, context, configSources);
+                        tryClassPath(uri, context, configSources);
+                    } else if (uri.getScheme().equals("jar")) {
+                        tryJar(uri, context, configSources);
+                    } else if (uri.getScheme().startsWith("http")) {
+                        tryHttpResource(uri, context, configSources);
+                    } else {
+                        throw ConfigMessages.msg.schemeNotSupported(uri.getScheme());
+                    }
+                }
+            }
+        }
+
+        return configSources;
+    }
+
+    private void tryFileSystem(final URI uri, final ConfigSourceContext context, final List<ConfigSource> configSources) {
+        final Path urlPath = Paths.get(uri.getPath());
+        if (Files.isRegularFile(urlPath)) {
+            consumeAsPath(toURL(urlPath.toUri()), new ConfigSourcePathConsumer(this, context, configSources));
```

Comment:
Why not this here:

```suggestion
            consumeAsPath(uri.getPath(), new ConfigSourcePathConsumer(this, context, configSources));
```

Reply:
If you're always passing `this` in as the first parameter of a nested class constructor, it probably should just be an inner class.

Reply:
> Why not this here:

Because we don't have a `consumeAsPath` in the `ClassPathUtils` that takes a `String`. We can probably add it. The other methods that do take a `String` handle it as a classpath resource and search the classloader. In this case, we just want to load this single resource.